text/0088-fix-memory-limitiations-in-session-replays-access-pattern.md

```diff
@@ -0,0 +1,414 @@
+
+### 1. Change the Product's Query Access Pattern
+
+**Proposal**
+
+Our current subquery solution works very well. However, there are escape hatches which require us to issue an aggregation query over the whole dataset. Those escape hatches should be closed.
+
+- Remove ability to filter and sort by aggregated values (e.g. count_errors, activity_score).
+- Remove ability to filter by negation values which change (e.g. urls).
+- Remove ability to use the AND operator when a non-static value is present.
+  - Filtering by browser_name AND os_name would be permitted.
+  - Filtering by browser_name AND click_action would be disallowed.
+    - Click rows do not store browser information because they are internally generated.
+    - Should this condition change these filters could be supported.
+  - Filtering by error_id AND url would be disallowed.
+    - Filtering by error_id OR url would be allowed.
+
+**Drawbacks**
+
+- Significant reduction in index page quality.
+- Removes the most important features of our index page. Count-errors sort, activity sort, duration sort and filter.
+
+**Questions**
+
+### 2. Reduce the Scan Range on the Front End
+
+**Proposal**
+
+The number of rows aggregated can be reduced by restricting the maximum time range we query over. We should validate the timestamp range such that it does not exceed a 24-hour period. This would satisfy every organization which ingests fewer than 1 billion replay-segments every 90 days.
+
+**Drawbacks**
+
+- To search for a unique value a user would need to issue a query for each day of the retention period or until the value was found.
+- Small customers would not see any replays on their index page due to the limited window size.
+  - Necessitates a special flag for large customers to enable this optimization.
+  - We may not know who needs this flag in advance and we may present a degraded customer experience without realizing.
+
+**Questions**
+
+### 3. Reduce the Scan Range on the Back End
+
+**Proposal**
+
+The number of rows aggregated can be reduced by restricting the maximum time range we query over. For select queries the backend can issue multiple queries on subsets of the range. For example, if we assume that no sort value was provided or that the sort value was applied to the timestamp column then the back end can transparently query a subset of the window attempting to populate the result set without querying the entire range.
+
+**Drawbacks**
+
+- Requires O(retention_days) queries to satisfy the result set in the worst case.
+- Adds an additional layer of complexity and does not solve our OOM issue.
+
+**Questions**
+
+### 4. Normalize Schema and Remove Snuba Join Restriction
+
+**Proposal**
+
+- Normalize schema.
+- Use joins.
+
+**Drawbacks**
+
+- Performance.
+- Requires Snuba work.
+
+**Questions**
+
+### 5. Migrate to CollapsingMergeTree Table Engine and Pre-Aggregate Data in Key-Value Store
```

Comment:
Infra Burden: Medium
```suggestion
### 5. Make a replays table as opposed to a replay_segments table
```
Infra burden: High
Customer Impact: Low
Maintainability: High

Notes: If we have a `replays` table instead of a `replay_segments` table we can filter better even if we aggregate the same amount of replays